### PR TITLE
CKBox image editing

### PR DESCRIPTION
**Pull request by arkflpc**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (ckbox): Integrates CKBox Image Editing feature with the editor. 

See https://github.com/cksource/ckeditor5-commercial/issues/5603.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-ckbox/docs/_snippets/features/ckbox.html [hunk 1]
```diff
@@ -15,4 +15,6 @@ <h2>Uploading non-embeddable files</h2>
 		<li>Select a PDF and click the <strong>Choose</strong> button.</li>
 	</ol>
 	<p>A link to the file will appear in the content.</p>
+	<h2>Editing images</h2>
+	<p>Use the image contextual toolbar to invoke the CKBoex editing pane right from the editor.</p>
```
Comment (arkflpc) [inline packages/ckeditor5-ckbox/docs/_snippets/features/ckbox.html:19]:
```suggestion
	<p>Use the image contextual toolbar to invoke the CKBox editing pane right from the editor.</p>
```

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* globals AbortController, URL, window */
```
Comment (arkflpc) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:6]: `utils/global`?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		} else {
+			this.isEnabled = false;
+		}
+	}
+
+	/**
+	 * Opens the CKBox Image Editor dialog for editing the image.
+	 */
+	public override execute(): void {
+		if ( !this.isEnabled || this._getValue() ) {
```
Comment (arkflpc) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:84]: Isn't `isEnabled` handled by `Command` class?

**Review by gorzelinski — COMMENTED** [review]
LGTM

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+		this._showImageProcessingIndicator( asset );
+
+		this._waitForAssetProcessed( asset.data.id, state.controller.signal )
+			.then(
+				asset => {
+					this._replaceImage( state.element, asset );
+				},
+				() => {
+					if ( !state.controller.signal.aborted ) {
+						notification.showWarning( t( 'Server failed image processing or didn\'t respond' ), {
```
Comment (arkflpc) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:208]: Its context is missing in `lang/contexts.json`.
@godai78 , is this sentence understandable?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+							title: t( 'Server failed image processing or didn\'t respond' ),
```
Comment (arkflpc) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:209]: I would invent something shorter for message title. This one just repeats the message.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 4 above)
Comment (godai78) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:208]: suggest `Server failed to process the image`.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+	}
+
+	/**
+	 * Save edited image. In case server respond with "success" replace with edited image,
+	 * otherwise show notification error.
+	 */
+	private _handleImageEditorOnSave( state: ProcessingState, asset: CKBoxRawAssetDefinition ) {
+		const t = this.editor.locale.t;
+		const notification = this.editor.plugins.get( Notification );
+		const pendingActions = this.editor.plugins.get( PendingActions );
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:195]: The `Notification` plugin should be included in the `CKBoxImageEditEditing` requires block.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+		this._processInProgress.set( state.ckboxImageId, state );
+		this._showImageProcessingIndicator( asset );
+
+		this._waitForAssetProcessed( asset.data.id, state.controller.signal )
+			.then(
+				asset => {
+					this._replaceImage( state.element, asset );
+				},
+				() => {
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:206]: This should re-throw (or log to the console) an exception if it's not `CKEditorError`. Silencing runtime exceptions could lead to unknown exceptions in runtime. For example bug in the code might be not noticed.

https://github.com/ckeditor/ckeditor5/blob/76cb59a2c6fe056c05551917f77e6c9c5a06fe4a/packages/ckeditor5-core/src/editor/editor.ts#L534-L538

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+
+					// Remove processing indicator. It was added only to ViewElement.
+					this.editor.editing.reconvertItem( state.element );
+				}
+			).finally( () => {
+				this._processInProgress.delete( state.ckboxImageId );
+				pendingActions.remove( action );
+			} );
+
+		this.refresh();
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:222]: Is this needed?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+		const ckboxEditing = this.editor.plugins.get( CKBoxEditing );
+		const response: CKBoxRawAssetDataDefinition = await sendHttpRequest( {
+			url,
+			signal,
+			authorization: ckboxEditing.getToken().value
+		} );
+		const status = response.metadata!.metadataProcessingStatus;
+
+		if ( !status || status == 'queued' ) {
+			throw new Error( 'Image has not been processed yet.' );
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:240]: We should use the `CKEditorError` to be able to distinguish runtime exceptions from known use cases.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+		}
+
+		return { data: { ...response } };
+	}
+
+	private async _waitForAssetProcessed( id: string, signal: AbortSignal ): Promise<CKBoxRawAssetDefinition> {
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:246]: Missing docs.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+		const result = await retry(
+			() => this._getAssetStatusFromServer( id, signal ),
+			{
+				signal,
+				maxAttempts: 5
+			}
+		);
+
+		if ( result.data.metadata!.metadataProcessingStatus != 'success' ) {
+			throw new Error( 'Image processing failed.' );
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:256]: We should use the `CKEditorError` to be able to distinguish runtime exceptions from known use cases.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+	}
+
+	/**
+	 * Shows processing indicator while image is processing.
+	 *
+	 * @param asset Data about certain asset.
+	 */
+	private _showImageProcessingIndicator( asset: CKBoxRawAssetDefinition ): void {
+		const editor = this.editor;
+		const selectedImageElement = editor.model.document.selection.getSelectedElement()!;
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:269]: Please use the `element` from `state`.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+
+		editor.editing.view.change( writer => {
+			const imageElementView = editor.editing.mapper.toViewElement( selectedImageElement )!;
+			const imageUtils: ImageUtils = this.editor.plugins.get( 'ImageUtils' );
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:273]: Is the `ImageUtils` in the `requires` block?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+			writer.addClass( 'image-processing', imageElementView );
+		} );
+	}
+
+	/**
+	 * Replace the edited image with the new one.
+	 */
+	private _replaceImage( element: ModelElement, asset: CKBoxRawAssetDefinition ) {
+		const editor = this.editor;
+		const imageCommand: InsertImageCommand = editor.commands.get( 'insertImage' )!;
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:292]: Is this command available? Are all `requires` plugins included?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+			imageHeight,
+			imagePlaceholder
+		} = prepareImageAssetAttributes( asset );
+
+		const previousSelectionRanges = Array.from( editor.model.document.selection.getRanges() );
+
+		editor.model.change( writer => {
+			writer.setSelection( element, 'on' );
+
+			imageCommand.execute( {
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:307]:
```suggestion
			editor.execute( 'insertImage', {
```

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+					...( imagePlaceholder ? { placeholder: imagePlaceholder } : null )
+				}
+			} );
+
+			const previousChildren = element.getChildren();
+
+			element = editor.model.document.selection.getSelectedElement()!;
+
+			for ( const child of previousChildren ) {
+				writer.append( child, element );
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:323]: This is modifying the `$graveyard`. You must not modify the content of the graveyard. After undo the original caption could be lost or lead to an unexpected OT error. Maybe we could use `writer.cloneElement( child )`.

**Review by mmotyczynska — CHANGES_REQUESTED** [review]
Only micro changes required.

Besides that, the `ckeditor5-metadata.json` file should be updated (e.g. add new icon).

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-ckbox/docs/features/ckbox.md [hunk 17]
```diff
@@ -30,16 +30,16 @@ As a full-fledged file manager, CKBox also replaces the basic CKEditor&nbsp;5 im
 * Enables drag & drop uploads of images and other files.
 * Transforms the _Image_ toolbar button, allowing the user to quickly upload and insert an image without opening the CKBox UI.
 * Adds a separate dedicated toolbar button to open the CKBox UI to manage and reuse uploaded files.
-* Provides basic editing capabilities, like crop, resize, rotate, and flip.
+* Provides {@link @ckbox features/images/editing basic editing capabilities}, like crop, resize, rotate, and flip right from the {@link features/images-overview#image-contextual-toolbar image contextual toolbar} for images served by CKBox.
```
[the remark this answers is not in the document]
Comment (godai78) [reply packages/ckeditor5-ckbox/docs/features/ckbox.md:33]: Local links to outside stuff (like CKBox) do not generate correctly. So this is not an actual error.

File: packages/ckeditor5-ckbox/src/utils.ts [hunk 18]
```diff
@@ -116,3 +116,80 @@ export function blurHashToDataUrl( hash?: string ): string | undefined {
...
+	url: URL;
+	signal: AbortSignal;
+	authorization: string;
+	method?: 'GET' | 'POST';
+	data?: FormData | null;
+	onUploadProgress?: ( evt: ProgressEvent ) => void;
+} ): Promise<any> {
+	const xhr = new XMLHttpRequest();
+
+	xhr.open( method, url.toString(), true );
```
[the remark this answers is not in the document]
Comment (gorzelinski) [reply packages/ckeditor5-ckbox/src/utils.ts:146]: Do we want to delete both arguments? It seems the [method](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open) requires at least two. TypeScript after the change yells at me.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 12 above)
Comment (gorzelinski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:272]: I used it, and unfortunately, it caused some tests to fail (`ckboximageedit/ckboximageeditcommand`).
Screenshot 2023-11-20 at 16 44 14

File: packages/ckeditor5-ckbox/src/utils.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-ckbox/src/utils.ts]: My mistake. What I meant was `xhr.open( method, url );`

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 11 above)
Comment (niegowski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:267]: I mean the case of `Uncaught TypeError: Cannot read properties of undefined (reading 'foo')` should not be displayed as a notification. It should be logged to the console as an error for a developer to fix. It's a code bug, not a known use case that the user should be informed of and would be able to solve.

**Comment by godai78** [comment]
Considering the scope of this PR, there might later be some conflicts with the Insert Image PR.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 12 above)
Comment (gorzelinski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts]: Update: I was able to make the changes and fix the tests.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 8 above)
Comment (gorzelinski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:224]: We were discussing this with @mmotyczynska  and it's probably required.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 13 above)
Comment (gorzelinski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:283]: When I add ImageUtils to the `requires` block, the linter tells me to move the `@ckeditor/ckeditor5-image` package to dependencies. Should I?

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 19]
```diff
@@ -205,8 +207,8 @@ export default class CKBoxImageEditCommand extends Command {
 				},
 				() => {
 					if ( !state.controller.signal.aborted ) {
-						notification.showWarning( t( 'Server failed image processing or didn\'t respond' ), {
-							title: t( 'Server failed image processing or didn\'t respond' ),
+						notification.showWarning( t( 'Server failed to process the image.' ), {
+							title: t( 'Image processing failed.' ),
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:211]: This message is not aligned with the `context.json`.

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+
+	/**
+	 * @inheritDoc
+	 */
+	public override refresh(): void {
+		const editor = this.editor;
+
+		this.value = this._getValue();
+
+		const selectedElement = editor.model.document.selection.getSelectedElement();
```
Comment (niegowski) [inline packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:65]: This should handle selection inside the image caption element.

https://github.com/ckeditor/ckeditor5/blob/490e44e0013a4016601f8484ef27d517d1e71ef0/packages/ckeditor5-image/src/imagetextalternative/imagetextalternativecommand.ts#L30-L31

**Review by niegowski — APPROVED** [review]

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts (hunk 20 above)
Comment (niegowski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:65]: Extracted to follow-up: https://github.com/cksource/ckeditor5-commercial/issues/5741

File: packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+			imageHeight,
+			imagePlaceholder
+		} = prepareImageAssetAttributes( asset );
+
+		const previousSelectionRanges = Array.from( editor.model.document.selection.getRanges() );
+
+		editor.model.change( writer => {
+			writer.setSelection( element, 'on' );
+
+			editor.execute( 'insertImage', {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts:323]: Extracted to follow-up: https://github.com/cksource/ckeditor5-commercial/issues/5742.

**Pull request merged**